docs/decisions/0003-lti-1p3-score-linking.rst

```diff
@@ -0,0 +1,94 @@
+
+
+Decisions
+=========
+
+We want to enable the platform to have full LTI Advantage compatibility, so we need to allow both interaction models to happen.
+
+To do that, a configuration option will be available to course creators when adding a new graded LTI 1.3 block to a course.
+This configuration will be called *Tool Grading Configuration* and offer the following options:
+
+.. list-table::
+   :widths: auto
+   :header-rows: 1
+
+   * - Option name
+     - Behavior
+   * - Only allow tools to push a single grade per student (Default)
+     - This is the "declarative" approach, where the platform will create a single LineItem, and only allow read-only access to it.
+       The tool is only able to read LineItems, push scores and retrieve grades.
+   * - Allow tools to manage their own grades
+     - This is the "programmatic" approach, where tools have full permissions to create, edit and delete LineItems, and well as
+       pushing and retrieving grades. Note that there might be cases where a tool doesn't set a `resouceLinkId` leaving the grade
+       unlinked to a problem. Also, the other edge case is when a tool pushes multiple grades for a single problem, in which case
+       the implementation needs to decide on a criteria to merge the grades or pick one.
+
+Declarative grade handling
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+When the LTI configuration is created, a signal should trigger the creation of a LineItem if the tool is configured to use the declarative
+model. The LineItem created will have the following attributes:
+
+.. list-table::
+   :widths: auto
+   :header-rows: 1
+
+   * - Attribute
+     - Value
+   * - lti_configuration
+     - LTI configuration just created.
+   * - resource_id
+     - Blank, this is used by LTI tools in the programmatic interaction model.
+   * - label
+     - The problem title, derived from the block's attributes.
+   * - score_maximum
+     - Maximum score for this given problem, derived from the block's attributes.
+   * - tag
+     - Blank, this is used by LTI tools in the programmatic interaction model.
+   * - start_date_time and end_date_time
+     - The problem's start and end date, if available in the block's attributes.
+
+Programmatic grade handling
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+No LineItems are created when the LTI configuration model is created, but the tools have permissions to create, update and
+delete LineItems using the LineItem endpoint.
+
+A *post_save* Django signal in the *LtiAgsScore* should be responsible for loading the XBlock from the modulestore,
+bind the user to the session, and set the score (after doing the proper scaling using the `scoreMaximum` attribute).
+
+If a tool creates and links multiple problems to the same grade, the platform will ??? the results.
+
+???: Not sure what should be the behavior here:
+1. Link just the latest grade submitted by the tool.
+2. Average all items and submit average to block.
```

Comment:
@nedbat This is my main concern regarding grade submission in the `programmatic` approach. Not sure what's the best option here. (1) is simpler, (2) doesn't leave any grading out of the final grade.

Reply:
IIUC, we are talking about multiple line items for the same problem, so conceptually, two different grades that are both "current" or "latest".  So I don't think #1 is an option: time-based "latest" isn't a way to choose between two grades that conceptually are side-by-side.  Or have I misunderstood?

Reply:
@nedbat I've assumed that most tools would just send a single grade, and in that case it won't be an issue.
I'll go with option 2 then, average all linked LineItems and push them the problem's grade.

Reply:
@marcotuts suggests that we only use "the primary grade."  Is there one that is clearly the primary?

Reply:
If I read the [`Score Service` spec](https://www.imsglobal.org/spec/lti-ags/v2p0#timestamp) correctly, assuming a score is submitted for a specific `LineItem` with a `user_id` field that already has a score recorded for it, the new score should overwrite the existing one (up to and including setting the score to blank and/or deleting it) as long as the timestamp is greater than the one already recorded.

Reply:
@pcockwell The issue is when there are multiple `LineItems` for a single block. In `edx-platform` a block only accepts a single grade.

We're discussing averaging the values retrieved by the tool. 

I think that the main issue here is that we don't have a good enough sample of LTI 1.3 to decide on an approach.